### PR TITLE
Support Figure of eight loitering

### PR DESCRIPTION
**Pull request by KonradRudin**
### Solved Problem
Add support for a figure of eight loitering pattern for fixed wing vehicles (can e.g. be used to replace orbit when tracking POIs with a limited panning gimbal). This adds the support for the MAV_CMD_DO_FIGURE_EIGHT mavlink command currently defined in the development dialect.

### Solution
- Add a new functionality in the FixedWingPositionControl to process a figure of eight loitering pattern. The class checks the relevant segment of the figure of eight and passes the path segment to the npfg guidance logic for lateral guidance.
- Addded a configuration to enable the figure of 8 loitering. It is disabled by default since it would need the development mavlink dialect (only enabled for sitl config).

### Changelog Entry
For release notes:
```
Feature Support for Figure 8 loiter pattern
```

### Test coverage
- Simulation tests performed in sitl.
- Build tests for fmu_v5x

Screenshot from 2023-07-14 16-50-24

### Context
To discuss:

- [ ] Currently QGCS has no support for sending this command only AMC is this a problem?
- [ ] Added a kconfig parameter to disable it since we need the development mavlink dialect to enable it. But i have not found a solution to make a dependency on the development dialect. Any input on how to handle this is appreciated.

### CODE REVIEW

**Comment by tstastny** [comment]
@dagar just so that this one doesnt get kicked down the road too far:

> Added a kconfig parameter to disable it since we need the development mavlink dialect to enable it. But i have not found a solution to make a dependency on the development dialect. Any input on how to handle this is appreciated.

do you have insights here how to handle? would the current action look something like default kconfiging the fig8 as disabled while there is a development mavlink link? (but then need to still make sure we have mavsdk test and a build that can run in ci.. even though on all the main binary builds it would be disabled)

> no GCS implementation

Is this already a nogo? or would mavsdk test be enough while this is in mavlink development.xml?

If the lack of operator command ease marks this as not good for merging now.. let's just decide quickly so that we'll take this back downstream until such point that we can bring it in more properly upstream.


@KonradRudin if we proceed with this PR - you may need to check the implications of this PR on the figure eight logic #21988

**Comment by Jaeyoung-Lim** [comment]
> Is this already a nogo? or would mavsdk test be enough while this is in mavlink development.xml?

@tstastny Why not just add it to QGC :smirk:?

**Comment by tstastny** [comment]
> @tstastny Why not just add it to QGC 😏?

@KonradRudin @RomanBapst  ^^

File: src/modules/mavlink/mavlink_messages.cpp [hunk 1]
```diff
@@ -120,6 +120,9 @@
 #include "streams/VFR_HUD.hpp"
 #include "streams/VIBRATION.hpp"
 #include "streams/WIND_COV.hpp"
+#ifdef CONFIG_FIGURE_OF_EIGHT
```
Comment (dagar) [inline src/modules/mavlink/mavlink_messages.cpp:123]: This part could be 
```suggestion
#if defined(MAVLINK_MSG_ID_FIGURE_EIGHT_EXECUTION_STATUS)
```

File: boards/px4/sitl/default.px4board [hunk 2]
```diff
@@ -18,6 +18,7 @@ CONFIG_MODULES_FLIGHT_MODE_MANAGER=y
 CONFIG_MODULES_FW_ATT_CONTROL=y
 CONFIG_MODULES_FW_AUTOTUNE_ATTITUDE_CONTROL=y
 CONFIG_MODULES_FW_POS_CONTROL=y
+CONFIG_FIGURE_OF_EIGHT=y
```
Comment (dagar) [inline boards/px4/sitl/default.px4board:21]: How about we make the mavlink part dependent on `MAVLINK_MSG_ID_ACTUATOR_CONTROL_TARGET`

Then either drop CONFIG_FIGURE_OF_EIGHT or make it clear it's FW (with a dependency on CONFIG_MODULES_FW_POS_CONTROL).

Comment (tstastny) [reply boards/px4/sitl/default.px4board:21]: > How about we make the mavlink part dependent on MAVLINK_MSG_ID_ACTUATOR_CONTROL_TARGET

did you mean `MAVLINK_MSG_ID_FIGURE_EIGHT_EXECUTION_STATUS`?

Comment (KonradRudin) [reply boards/px4/sitl/default.px4board:21]: @dagar the dependency is already defined in the kconfig, i guess this should be enough?

Comment (dagar) [reply boards/px4/sitl/default.px4board:21]: It doesn't need to be completely tied to mavlink.

Comment (KonradRudin) [reply boards/px4/sitl/default.px4board:21]: Hey @dagar with the new commit, i have removed the dependency to the development dialect. you can now build the figure of eight with the common, you just wont get any feedback message.

**Review by tstastny — DISMISSED** [review]

**Comment by tstastny** [comment]
@dagar is this ok to go in now? before we e.g. run out of flash / need another rebase with other fixed-wing things on the docket?

@KonradRudin i re-started the tailsitter test, it was failing, but i didnt see an erroneous timeout.. may actually be something wrong, can you check? https://github.com/PX4/PX4-Autopilot/actions/runs/6655895059/job/18096246149#step:17:1383

**Comment by tstastny** [comment]
@dagar all checks passing - should we smash the button?

**Comment by KonradRudin** [comment]
@tstastny had to rebase, since it had some merge conflicts (only in the mavsdk tests).

**Comment by hamishwillee** [comment]
@KonradRudin A bit annoying this appeared without docs. That would explain why know one knows about it. Can you look at https://github.com/PX4/PX4-Autopilot/pull/26194 please.

**Pull request merged**